samza-core/src/main/java/org/apache/samza/classloader/IsolatingClassLoaderFactory.java

```diff
@@ -0,0 +1,316 @@
+   *
+   * Delegation will only happen for classes specified in
+   * {@link DependencyIsolationUtils#FRAMEWORK_API_CLASS_LIST_FILE_NAME} and the Java bootstrap classes.
+   */
+  private static DelegateRelationship buildApiParentRelationship(File apiLibDirectory, ClassLoader apiClassLoader) {
+    DelegateRelationshipBuilder apiParentRelationshipBuilder = DelegateRelationshipBuilder.builder()
+        // needs to load API classes from the API classloader
+        .withDelegateClassLoader(apiClassLoader)
+        /*
+         * FULL means to only load classes explicitly specified as "API" from the API classloader. We will use
+         * delegate-preferred class predicates to specify which classes are "API" (see below).
+         */
+        .withIsolationLevel(IsolationLevel.FULL);
+
+    // bootstrap classes need to be loaded from a common classloader
+    apiParentRelationshipBuilder.addDelegatePreferredClassPredicate(new BootstrapClassPredicate());
+    // the classes which are Samza framework API classes are added here
+    getFrameworkApiClassGlobs(apiLibDirectory).forEach(
+        apiClassName -> apiParentRelationshipBuilder.addDelegatePreferredClassPredicate(new GlobMatcher(apiClassName)));
+    return apiParentRelationshipBuilder.build();
+  }
+
+  /**
+   * @param directoryWithClassList Directory in which
+   * {@link DependencyIsolationUtils#FRAMEWORK_API_CLASS_LIST_FILE_NAME} lives
+   * @return {@link List} of globs for matching against classes to load from the framework API classloader
+   */
+  @VisibleForTesting
+  static List<String> getFrameworkApiClassGlobs(File directoryWithClassList) {
+    File parentPreferredFile =
+        new File(directoryWithClassList, DependencyIsolationUtils.FRAMEWORK_API_CLASS_LIST_FILE_NAME);
+    validateCanAccess(parentPreferredFile);
+    try {
+      return Files.readAllLines(Paths.get(parentPreferredFile.toURI()), StandardCharsets.UTF_8);
+    } catch (IOException e) {
+      throw new SamzaException("Error while reading samza-api class list", e);
+    }
+  }
+
+  /**
+   * Get the {@link URL}s of all JARs/WARs in the directory {@code jarsLocation}. This only looks one level down; it is
+   * not recursive.
+   */
+  @VisibleForTesting
+  static URL[] getClasspathAsURLs(File jarsLocation) {
+    validateCanAccess(jarsLocation);
+    File[] filesInJarsLocation = jarsLocation.listFiles();
+    if (filesInJarsLocation == null) {
+      throw new SamzaException(
+          String.format("Could not find any files inside %s, probably because it is not a directory",
+              jarsLocation.getPath()));
+    }
+    URL[] urls = Stream.of(filesInJarsLocation)
+        .filter(file -> file.getName().endsWith(".jar") || file.getName().endsWith(".war"))
+        .map(IsolatingClassLoaderFactory::fileURL)
+        .toArray(URL[]::new);
+    LOG.info("Found {} items to load into classpath from {}", urls.length, jarsLocation);
```

Comment:
Might be useful  for debugging to log actual directory contents, either here or in buildClassLoader().